Sprint 15

folawer21 opened the pull request.

folawer21 committed:
a

folawer21 committed:
надо добавить scroll view
+
добавить смайлики и цвет

folawer21 committed:
emoji added

folawer21 committed:
colors added
Need to add buttons + collectionView or take Scroll view with main collection + buttons

folawer21 committed:
Даты готовы!

folawer21 committed:
done event

folawer21 committed:
done 14

helecta commented on tracker/TrackerView/Cells/TrackCell.swift:
```diff
@@ -21,7 +26,14 @@ final class TrackCell: UICollectionViewCell{
     let plusButton = UIButton()
     var buttonWasTapped = false
     
+    let plusImage = UIImage(named: "plus")?.withRenderingMode(.alwaysTemplate)
```
Здесь стоит добавить private
Необходимо указывать private для того, чтобы спрятать детали реализации конкретной части функциональности,
когда они используются только внутри области объявления.

helecta commented on tracker/TrackerView/CreateEventVC.swift:
```diff
@@ -1,8 +1,156 @@
-//  tracker
 //
-//  Created by Александр  Сухинин on 29.05.2024.
+//import UIKit
```
обычно перед сдачей принято удалять закомментированный код и print’ы. Так код проще читать, а дебажная информация не должна попадать в продакшен(т.е print’ы). За исключением спринтов, где требуется логирование

helecta commented on tracker/TrackerView/CreateEventVC.swift:
```diff
@@ -103,15 +204,27 @@ extension CreateEventVC: UICollectionViewDataSource{
             guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "EventCell", for: indexPath) as? EventCell else {return UICollectionViewCell()}
-//            cell.delegate = self
+            return cell
+        case 2:
```
Здорово было бы вынести такие константы в enum
```
private enum Sections: Int, CaseIterable  {
    case name = 0
	  case event
    case buttons
    case emoji
    case color
}
```

Тогда можно было бы так сделать:
```
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return Sections.allCases.count
    }
```
И так:
```
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        guard let section = Sections(rawValue: indexPath.section) else {
            return UICollectionViewCell()
        }
        switch section {
        case Sections.name:
            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: TrackNameCell.identifier, for: indexPath) as? TrackNameCell else {
                return UICollectionViewCell()
            }
            cell.textFieldDelegate = self
            cell.prepareForReuse()
            return cell
            
        case Sections.buttons:
            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CreateCancelButtonsCells.identifier, for: indexPath) as? CreateCancelButtonsCells else {
                return UICollectionViewCell()
            }
            cell.setupView()
            cell.delegate = self
            return cell
        ...
```
Так читать проще, не нужно держать в голове последовательность секций.
Так же секции легко станет поменять местами, просто порядок в enum меняем и все.

helecta commented on tracker/TrackerView/CreateHabbitVC.swift:
```diff
@@ -145,39 +105,99 @@ extension CreateHabbitVC: UICollectionViewDelegateFlowLayout{
         return UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 0)
     }
     
+    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
```
обрати внимание, что у тебя код CreateHabbitVC очень похож на CreateEventVC. Обычно это повод для рефакторинга. Советую изучить тебе запахи кода, конкретно тут дублирование кода и как с этим бороться: https://refactoring.guru/ru/smells/duplicate-code

folawer21 committed:
CoreData including ...20%

folawer21 committed:
fixed days

folawer21 committed:
bag fixing

folawer21 committed:
before changing to CoreData

folawer21 committed:
coredataaaa

folawer21 committed:
core dataaaa

folawer21 committed:
УРРАААА Я СМОГ осталость трекерс рекордс
Достаточно было ударить по столу

folawer21 committed:
можно выбирать эмлдхи и цвет надо добавить блокировку кнопок пока они не выбраны

folawer21 committed:
color and emoji chosing is ready

folawer21 committed:
trackerRecordCd looks pretty done
commit before changing TrackerViewController and TrackerCell

need to connect datePicker to CD

folawer21 committed:
trackerRecordStore updated

folawer21 committed:
problems with reading from coredata

folawer21 committed:
done with all without recording

folawer21 committed:
add new functions 2 trackerRecordStore

folawer21 committed:
done

helecta commented on tracker/Stores/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,182 @@
+
+final class TrackerCategoryStore: NSObject{
+    private let manager = CDManager.shared
+    weak var trackerStore: TrackerStore?
```
Не забывай организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Подробнее почитать можно тут: [Организация кода](https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization)

helecta commented on tracker/Stores/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,200 @@
+            let count = try context.fetch(request).count
+            return count
+        }catch{
+            fatalError()
```
Обычно в продакшене не используют fatalError, советую заменить на assertionFailure, тут подробнее:
 [https://www.hackingwithswift.com/plus/intermediate-swift/understanding-assertions](https://www.hackingwithswift.com/plus/intermediate-swift/understanding-assertions)

helecta commented on tracker/TrackerView/Cells/ColorCells.swift:
```diff
@@ -0,0 +1,113 @@
+        cell.hideBlock()
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
```
Лучше такие длинные методы разбивать строчками:
```
            collectionView(_ collectionView: UICollectionView,
                layout collectionViewLayout: UICollectionViewLayout,
                sizeForItemAt indexPath: IndexPath) -> CGSize
```
Так код значительно легче и удобней читать

helecta commented on tracker/TrackerView/Cells/TrackCell.swift:
```diff
@@ -1,54 +1,206 @@
+            plusButton.layer.opacity = 0.3
+            enableButton()
+        }else{
+            if availible{
```
лучше стараться избегать вложенностей в if, усложняет код.
```
} else if availible {
            buttonWasTapped = false
            plusButton.isSelected = false
            enableButton()
} else {
            buttonWasTapped = false
            plusButton.isSelected = false
            plusButton.layer.opacity = 0.3
            disableButton()
}
```

helecta commented on tracker/TrackerView/Cells/TrackCell.swift:
```diff
@@ -1,54 +1,206 @@
+//                print("[buttonTapped] TrackCell - unable to get Id ")
+//                return
+//            }
+//            delegate?.addTrackerRecord(id: id)
```
обычно перед сдачей принято удалять закомментированный код и print’ы. Так код проще читать, а дебажная информация не должна попадать в продакшен(т.е print’ы). За исключением спринтов, где требуется логирование

helecta commented on tracker/TrackerView/CreateEventVC.swift:
```diff
@@ -103,15 +75,29 @@ extension CreateEventVC: UICollectionViewDataSource{
+            cell.delegate = self
+            cell.setupView()
+            return cell
+        case 3:
```
Здорово было бы вынести такие константы в enum
```
private enum Section: Int, CaseIterable  {
    case name = 0
    case event
    case buttons
    case emoji
    case color
}
```

Тогда можно было бы так сделать:
```
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return Section.allCases.count
    }
```

И так:
```
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        switch indexPath.section {
        case Section.name:
            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "TextField", for: indexPath) as? TrackNameCell else {return UICollectionViewCell()}
            cell.textFieldDelegate = self
            return cell
            
        case Section.buttons:
            ...
        case Section.emoji:
            ...
        case Sections.color:
            ...
        }
    }
```
Так читать проще, не нужно держать в голове последовательность секций.
Так же секции легко станет поменять местами, просто порядок в enum меняем и все.

helecta commented on tracker/TrackerView/CreateHabbitVC.swift:
```diff
@@ -23,93 +26,70 @@ final class CreateHabbitVC: UIViewController{
             guard let day = WeekDay(rawValue: arr[i]) else { break}
             result.append(day)
         }
+        if result == [] {
```
можно использовать `result.isEmpty`
